File: python/do.py
#!/usr/bin/env python
from optparse import OptionParser
import cjson

do_action_def = {
  "Name" : "Bizzle fo shizzle",
  "Description" : "Some pimp shit",
  "Icon" : "text-x-script",
  "SupportedItemTypes" : ["Do.Universe.ITextItem", "Do.Universe.IFileItem" ]
}

def main():                       
  parser = OptionParser()
  parser.add_option("-d", "--do-action-def", action="store_true", default=False)
  (options, args) = parser.parse_args()
	
  if options.do_action_def:
    print_do_action_def()
  else:
    run_with_items(args[0])

def print_do_action_def():
  print cjson.encode(do_action_def)

def run_with_items(json):
  items = cjson.decode(json)
  
  foo = ""

  for item in items:
    if "Text" in item.keys():
      foo = item["Text"]

  return_items = [{ 
    "__type" : "Do.Universe.TextItem",
    "Name" : "Python name",
    "Description" : "Python description",
    "Icon" : "text-x-script",
    "Text" : foo}
  ]
  
  print cjson.encode(return_items)


if __name__ == "__main__":
  main()

File: ruby/do.rb
#!/usr/bin/env ruby
require 'rubygems'
require 'json'
require 'trollop'

do_action_def = { 
  :Name => 'Get Current Time', 
  :Description => 'Gets the current time from a script.', 
  :Icon => 'text-x-script',
  :SupportedItemTypes => ['Do.Universe.ITextItem', 'Do.Universe.IFileItem', 'OpenSearch.IOpenSearchItem']
}

opts = Trollop::options do
  opt :do_action_def, "Print GNOME Do Action definition"
end

if opts[:do_action_def]
  puts do_action_def.to_json
else 
  items_json = ARGV[0]
  #items_json = File.read(ARGV[0])
  
  foo = ""
  items = JSON.parse(items_json)
  items.each do |item|
    if item.has_key? 'Text'
      foo = item['Text']
    end
  end
  
  return_items = [
    { :__type => 'Do.Universe.TextItem',
      :Name => 'Text From Script',
      :Description => 'Description From Script',
      :Icon => 'gnome-mime-text',
      :Text => foo
      }
    ]
    
  puts return_items.to_json
end

